### PR TITLE
[SPARK-5376][Mesos] MesosExecutor should have correct resources

### PR DESCRIPTION
**Pull request by jongyoul**
- Divided task and executor resources
- Added `spark.mesos.executor.cores` and fixed docs

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -341,6 +341,13 @@ Apart from these, the following properties are also available, and may be useful
     `spark.mesos.executor.memoryOverhead` or 7% of `spark.executor.memory`.
   </td>
 </tr>
+<tr>
+  <td><code>spark.mesos.executor.cpus</code></td>
+  <td>1</td>
+  <td>
+    This value is the amount of cores so that executor is running itself.
```
Comment (tnachen) [inline docs/configuration.md]: Spark code has been referring cpus to cores, so I assume you want to name this spark.mesos.executor.cores
I think we should re-word this, to something like:
The amount of cores to request for running the mesos executor.

File: docs/running-on-mesos.md [hunk 2]
```diff
@@ -226,6 +226,13 @@ See the [configuration page](configuration.html) for information on Spark config
     The final total amount of memory allocated is the maximum value between executor memory plus memoryOverhead, and overhead fraction (1.07) plus the executor memory.
   </td>
 </tr>
+<tr>
+  <td><code>spark.mesos.executor.cpus</code></td>
+  <td>1</td>
+  <td>
+    The amount of cores that Mesos executor will request additionally for running executor itself.
```
Comment (tnachen) [inline docs/running-on-mesos.md]: Please use the same description in the other doc, and the change the value too.

File: docs/configuration.md (hunk 1 above)
Comment (jongyoul) [inline docs/configuration.md]: I agree. I'm also confused between cpus and cores. I think `cores` is more proper term. Thanks.

**Comment by mateiz** [comment]
Sorry, what was the problem here? Executors *do* use memory, in fact they use all of the memory of that JVM (the only reason we assign memory to tasks is that Mesos didn't support tasks with 0 memory).

**Comment by jongyoul** [comment]
@mateiz We agree with one executor and multi task is intended behaviour. In this situation, MesosScheduler offers CPUS_PER_TASK resources to executor when we launch separate task. If we launch two tasks, we offers 3 * CPUS_PER_TASK(= 1 for executor and 2 for tasks) for running only two tasks. @pwendell thinks that's too much resources and It's enough for executor to have one cores. In my PR, I enable to set executor's cores. In memory, we just offered memory to executor only. If we launch two tasks again, we offers 1 * calculateTotalMemory(sc) to all tasks. I think we offer one executor memory and two task memories. I agree that executor uses memory by itself, but we fix amount of those value

**Comment by jongyoul** [comment]
Thus, I believe that executor has executor cores and executor memory setting on ExecutorInfo, and task has its own cores and memories setting on TaskInfo while launching per task.

**Comment by jongyoul** [comment]
@mateiz That's a sample screenshot.
screen shot 2015-01-28 at 10 55 07 am

**Comment by mateiz** [comment]
Sorry, I'm not sure I understand about the memory. There should be the same amount of memory for each executor. More executors means more total memory. But the way that was being calculated before is fine, what's wrong with it?

For CPUs, I understand wanting to start executors off at 0 CPUs and just having 1 per task. I actually thought we did that initially, not sure why it's 1.

**Comment by mateiz** [comment]
Basically I don't understand what your patch is fixing. What's an example of config settings that gave the wrong allocation before, and what will they give now?

**Comment by jongyoul** [comment]
Sorry, I've shaw you my configuration. my configuraion is 5G for SPARK_EXECUTOR_MEMORY and 5 for spark.task.cpus. In my screenshot, we launch two tasks on the same machine. Don't you think It's good to offer task memory twice? My PR gives correct resource management information to mesos' master. For CPUs, I don't know proper value of executor cpus, but not CPUS_TASK_CPUS. Recommend this value, please.

**Comment by mateiz** [comment]
Right, as I said, it doesn't make sense to offer task memory twice. Each executor is a *single* JVM, and JVMs cannot scale their memory up and down. The executor's memory is set to the same value that we configure that JVM with, with `-Xmx`. There's no way to make tasks use more memory than that, no matter how many tasks are running on there.

**Comment by tnachen** [comment]
@jongyul sorry didn't get to finish reviewing the PR, and I agree with matei that in spark usage of mesos it doesn't make sense to give tasks memory, as we share the same executor that is kept running.

**Comment by jongyoul** [comment]
I don't know the behaviour in coarse-grained mode, but in fine-grained mode, we use multiple JVM for running tasks. we run spark-class by launcher. This means we launch JVM by running per task. Am I wrong?

**Comment by jongyoul** [comment]
I believed that when we launch mesos driver launchTasks, container run the command `bin/spark-class` everytime running task. And in my qna email for mesos, @tnachen answers that one container run multiple command simultaneously. And my some tests show two tasks runs simutaneously because they write a same log file at the same time. And my digging codes results no limit to launch task on a mesos container. However, @mateiz told me that one executor only runs a single JVM and launch a single task at any time.

**Comment by tnachen** [comment]
If you read the fine-grained mode source code, you'll notice that Spark is using the slave id as the executor id, which is what we discussed on the mesos mailing list, that the executor will be re-used if all tasks reuse the same executor id.
Therefore, it's only launching one executor per slave, and if the executor dies Mesos will relaunch it when the task asks for it again.

**Comment by jongyoul** [comment]
@tnachen Yes, I fully understand reusing executor while a framework is alive. However, we launch two  task on a same executor? What you've answered is they are launched at the same time, isn't it?

**Comment by tnachen** [comment]
@jongyoul So an executor can only "launch" one task at a time, but can have multiple tasks running simultaneously as you mentioned.

It doesn't matter if they're all part of the same launchTasks message or seperate, as long as the framework and executor id are the same it will be launched in the same executor.

**Comment by jongyoul** [comment]
@tnachen @mateiz So sorry for taking up a lot of time. I've found that only one executor as a process runs at any time, and I understand executor can have multiple tasks at the same time. I've believed each executor is launched separately when driver launchTasks.

**Comment by jongyoul** [comment]
I'll close this PR. It's wrong approach.

**Comment by elyast** [comment]
One comment, however if you run multiple Spark applications even tough executor-id == slave-id, multiple executors can be started on the same host. (And every one of them will consume 1 CPU without scheduling any tasks). This can be painful when you want to run multiple streaming applications on Mesos in fine grained mode, because each streaming driver's executors will consume 1 CPU...


executors

Screen shots illustrate situation on single slave, when there are two executors running for 2 different Spark applications (one is streaming app, second one is Zeppelin), and as u can see there 0 active tasks the consumption of CPU is 2.

**Comment by tnachen** [comment]
@elyast yes you are correct it is only applicable per Spark app. It is entirely possible to make executor cpu less than 1 (as it's based on shares), but it's not possible for now to share mesos executor across apps.

**Comment by elyast** [comment]
Sure its totally fine not to share, but at least it should be possible to configure allocation. Allocating 1 CPU per executor may just too much, obviously it depends how cpu intensive is his work, but I guess @mateiz know that much better than me

**Comment by jongyoul** [comment]
@elyast Thanks for interesting this PR, which was about resources of cores and memory. I misunderstood how mesos works specially in memory side, so I closed this PR. However, I agree with you that executor cores are sometime too much. @tnachen we cannot adjust memory issue, but, changing executor cores is meaningful and should fix that executor has same cores with CPUS_PER_TASK initially. See TODO(pwendell) in MesosSchedulerBackend.scala. I want to fix this TODO to extract a configuration parameter. What do you think of it?

**Comment by tnachen** [comment]
I think making it a configurable parameter sounda reasonable to me.

**Comment by jongyoul** [comment]
@tnachen @elyast I made a new issue about configuring mesos executor cores. https://issues.apache.org/jira/browse/SPARK-6350

**Comment by elyast** [comment]
cool thanks

**Pull request closed without merging**